pkg/entities/tokens.go

```diff
@@ -201,3 +206,37 @@ func (e *Entity) TotalSupportedTokens(guildId string) (*response.Metric, error)
 func (e *Entity) GetTokensByChainID(chainID int) ([]model.Token, error) {
 	return e.repo.Token.GetByChainID(chainID)
 }
+
+func (e *Entity) GetTokenBalanceFunc(chainID, address string) (func(address string) (*big.Int, error), error) {
+	chainIDNum, err := strconv.Atoi(chainID)
+	if err != nil {
+		e.log.Errorf(err, "[strconv.Atoi] failed to convert chain %s to number", chainID)
+		return nil, fmt.Errorf("failed to convert string to int: %v", err)
+	}
+	chain, err := e.repo.Chain.GetByID(chainIDNum)
+	if err != nil {
+		e.log.Errorf(err, "[repo.Chain.GetByID] failed to get chain by id %s", chainID)
+		return nil, fmt.Errorf("failed to get chain by id %s: %v", chainID, err)
+	}
+
+	// TODO: Support non evm
+	client, err := ethclient.Dial(chain.RPC)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to chain client: %v", err.Error())
+	}
+
+	contractERC20, err := erc20.NewErc20(common.HexToAddress(address), client)
+	if err != nil {
+		e.log.Errorf(err, "[erc20.NewErc20] failed to init erc20 contract")
+		return nil, fmt.Errorf("failed to init erc20 contract: %v", err.Error())
+	}
+	balanceOf := func(address string) (*big.Int, error) {
+		b, err := contractERC20.BalanceOf(nil, common.HexToAddress(address))
```

Comment:
can reuse func from `chain.go`?

Reply:
The version from chain.go using float64 for balance number, and the db schema is storing the required amount in numeric type (big.Int value) which I believe would mismatch the type 🤔